### ISSUE TITLE
Explain output contains redundancies

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#5076]
The `explain` output can contain noisy redundant expressions that don't really exist in the `RelationExpr`.

For example, TPCH Q04's explain output is
```
                          Optimized Plan                           
-------------------------------------------------------------------
 %0 =                                                             +
 | Get materialize.public.orders (u17)                            +
 | Filter (datetots(#4) < 1993-10-01 00:00:00), (#4 >= 1993-07-01)+
                                                                  +
 %1 =                                                             +
 | Get %0                                                         +
                                                                  +
 %2 =                                                             +
 | Get %0                                                         +
 | Filter !(isnull(#0))                                           +
 | Distinct group=(#0)                                            +
 | ArrangeBy (#0)                                                 +
                                                                  +
 %3 =                                                             +
 | Get materialize.public.lineitem (u7)                           +
 | Filter !(isnull(#0)), (#11 < #12)                              +
                                                                  +
 %4 =                                                             +
 | Join %2 %3 (= #0 #1)                                           +
 | | implementation = Differential %3 %2.(#0)                     +
 | | demand = (#0)                                                +
 | Distinct group=(#0)                                            +
 | ArrangeBy (#0)                                                 +
                                                                  +
 %5 =                                                             +
 | Join %1 %4 (= #0 #9)                                           +
 | | implementation = Differential %1 %4.(#0)                     +
 | | demand = (#5)                                                +
 | Reduce group=(#5)                                              +
 | | agg count(true)                                              +
                                                                  +
 Finish order_by=(#0 asc) limit=none offset=0 project=(#0, #1)    +
```
Up above, notice the
```
 %1 =                                                             +
 | Get %0                                                         +
```
which is clearly not very meaningful; each instance of `%1` could be replaced by `%0`.

What seems to be happening is that in `explain.rs`, we create a basic block for each input to a `Join` or `Union`, so that they can reference something when announcing their inputs. In some cases, those inputs may just be `Get { id: LocalId(_), .. }` and we could use the name associated with the let binding. Instead, we create a basic block containing just the `Get`, and use that name, as seen above.

I'm not really sure about the right "area" for this. The code is in the `expr` crate, but it is about user-facing stuff we document as "SQL".

**Comment by benesch** [comment]
I think this is "working as intended" though I also find it confusing. At the very least it would be nice to indicate which blocks were actual lets and which were synthesized by the explain output. I also didn't mind the nested output nearly as much as Jamie did. Deep nesting is definitely hard to follow, but the nested output did have the upside that an individual join or union with multiple inputs was represented very naturally.

**Comment by frankmcsherry** [comment]
This format works great for me, fwiw! Mostly, it seemed that the 
```
 %1 =                                                             +
 | Get %0                                                         +
```
stuff reflected some sort of issue on our part (could have been bad plans, for example, but fortunately wasn't). I could imagine "working as intended" in the sense that "it's not like there's a spec", but it does seem like a thing that could be improved making the cosmetics and confidence a bit better.

Or did you mean "I think that fragment should stay in, as it is intended to be there"?

**Comment by benesch** [comment]
Ah, I just meant that the philosophy of the current format seems to prioritize "flatness" above "faithfulness to source expression". With the old format there was a 100% chance you'd be able to reconstruct the exact input expression, but with this format we seem to have taken a minor hit on that front to avoid the nesting.

Anyway, I'm assuming you were thinking something like this, yeah?

```
%0 = Let l0 =
| Get materialize.public.orders (u18)
| Filter (datetots(#4) < 1993-10-01 00:00:00), (#4 >= 1993-07-01)

%1 =
| Get %0 (l0)
| ArrangeBy (#0)

%2 =
| Get materialize.public.lineitem (u21)
| ArrangeBy (#0)

%3 =
| Join %1 %2 (= #0 #9)
| | implementation = DeltaQuery
| |   delta %1 %2.(#0)
| |   delta %2 %1.(#0)
| | demand = (#0, #20, #21)
| Filter (#20 < #21)
| Distinct group=(#0)
| ArrangeBy (#0)

%4 =
| Join %0 %3 (= #0 #9)
| | implementation = Differential %0 %3.(#0)
| | demand = (#5)
| Reduce group=(#5)
| | agg count(true)

Finish order_by=(#0 asc) limit=none offset=0 project=(#0, #1)
```
